src/sdam/server_selection_events.ts

```diff
@@ -0,0 +1,137 @@
+import {
+  SERVER_SELECTION_FAILED,
+  SERVER_SELECTION_STARTED,
+  SERVER_SELECTION_SUCCEEDED,
+  WAITING_FOR_SUITABLE_SERVER
+} from '../constants';
+import { type ReadPreference } from '../read_preference';
+import { type ServerSelector } from './server_selection';
+import type { TopologyDescription } from './topology_description';
+
+/**
+ * The base export class for all logs published from server selection
+ * @internal
+ * @category Log Type
+ */
+export abstract class ServerSelectionEvent {
+  /** String representation of the selector being used to select the server.
+   *  Defaults to 'custom selector' for application-provided custom selector case.
+   */
+  selector: string | ReadPreference | ServerSelector;
+  /** The name of the operation for which a server is being selected.  */
+  operation: string;
+  /** 	The current topology description.  */
+  topologyDescription: TopologyDescription;
+
+  /** @internal */
+  abstract name:
+    | typeof SERVER_SELECTION_STARTED
+    | typeof SERVER_SELECTION_SUCCEEDED
+    | typeof SERVER_SELECTION_FAILED
+    | typeof WAITING_FOR_SUITABLE_SERVER;
+
+  abstract message: string;
+
+  /** @internal */
+  constructor(
+    selector: string | ReadPreference | ServerSelector,
+    topologyDescription: TopologyDescription,
+    operation?: string
+  ) {
+    this.selector = selector;
+    this.operation = operation ?? 'n/a';
```

Comment:
Can server selection happen without operation that caused it?

Reply:
Not that I know of, I'll make it a non-optional property.